File: Day 8/solution.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>
#include <unordered_map>
#include <functional>

constexpr auto max_size = std::numeric_limits<std::streamsize>::max();

class Solver
{
	std::filesystem::path filepath;

	std::pair<std::string, std::unordered_map<std::string, std::pair<std::string, std::string>>> parse_order_and_graph()
	{
		std::ifstream file;
		file.open(this->filepath);

		char c;
		// Get order
		std::string order;
		while ((c = file.get()) != '\n')
		{
			order += c;
		}

		// Blank line
		file.ignore(max_size, '\n');

		// Graph
		std::unordered_map<std::string, std::pair<std::string, std::string>> graph;
		while (file.peek() != EOF)
		{
			// Source node
			std::string node, left, right;
			for (c = file.get(); c != ' '; c = file.get())
			{
				node += c;
			}

			// Skip to left node
			file.ignore(max_size, '(');

			// Left node
			for (c = file.get(); c != ','; c = file.get())
			{
				left += c;
			}

			// Skip to right node
			for (c = file.get(); c == ' '; c = file.get())
			{
			}

			// Right node
			for (; c != ')'; c = file.get())
			{
				right += c;
			}

			// Skip to next line
			file.ignore(max_size, '\n');

			graph[node] = std::make_pair(left, right);
		}
		file.close();

		return std::make_pair(order, graph);
	}

	long get_steps(const std::string &order, std::string node, const std::unordered_map<std::string, std::pair<std::string, std::string>> &graph, std::function<bool(std::string)> is_end_node)
	{
		long steps = 0;
		while (!is_end_node(node))
		{
			char direction = order[steps++ % order.size()];
			node = direction == 'L' ? graph.at(node).first : graph.at(node).second;
		}
		return steps;
	}

	long gcd(long a, long b)
	{
		if (a < b)
			std::swap(a, b);

		while (b != 0)
		{
			long temp = a % b;
			a = b;
			b = temp;
		}
		return a;
	}

	long lcm(std::vector<long> nums)
	{
		while (nums.size() > 1)
		{
			long a = nums.back(), b;
			nums.pop_back();
			b = nums.back();
			nums.pop_back();
			nums.push_back(a / this->gcd(a, b) * b);
		}
		return nums.front();
	}

	long part_1()
	{
		auto [order, graph] = this->parse_order_and_graph();

		std::string start_node = "AAA";
		if (!graph.contains(start_node))
			return -1;

		return this->get_steps(order, start_node, graph, [](const std::string &node)
													 { return node == "ZZZ"; });
	}

	long part_2()
	{
		auto [order, graph] = this->parse_order_and_graph();
		std::vector<long> cycles;
		for (auto [node, _] : graph)
		{
			if (node.back() != 'A')
				continue;
			cycles.push_back(this->get_steps(order, node, graph, [](const std::string &node)
																			 { return node.back() == 'Z'; }));
		}

		return this->lcm(cycles);
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 8" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}

File: Day 8/solution.py
"""
Day 8 solution.
"""
import itertools
import sys
from typing import Callable


class Solver:
    """Day 8 solver."""

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    def parse_order_and_graph(self) -> tuple[str, dict[str, tuple[str, str]]]:
        """
        Parse the file to get the order and build the graph.
        """
        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            order = file.readline().strip()
            file.readline()  # Blank lines

            graph = {
                node: tuple(options.strip()[1:-1].split(", "))
                for node, options in (line.split(" = ") for line in file)
            }

            return order, graph

    @staticmethod
    def get_steps(
        order: str,
        node: str,
        graph: dict[str, tuple[str, str]],
        is_end_node: Callable[[str], bool],
    ) -> int:
        """Count the number of steps to reach the end node."""
        steps = 0
        for direction in itertools.cycle(order):
            node = graph[node][0 if direction == "L" else 1]
            steps += 1
            if is_end_node(node):
                return steps
        return -1

    @staticmethod
    def gcd(a: int, b: int) -> int:
        """The greatest common divisor."""
        if a < b:
            a, b = b, a

        while b != 0:
            a, b = b, a % b
        return a

    @staticmethod
    def lcm(nums: list[int]) -> int:
        """The lowest common multiple."""
        while len(nums) > 1:
            a, b = nums.pop(), nums.pop()
            nums.append(a * b // Solver.gcd(a, b))
        return nums[0]

    def part_1(self) -> int:
        """Part 1 solver."""
        order, graph = self.parse_order_and_graph()

        start_node = "AAA"
        if start_node not in graph:
            return -1

        return self.get_steps(
            order, start_node, graph, lambda node: node == "ZZZ"
        )

    def part_2(self) -> int:
        """Part 2 solver."""
        order, graph = self.parse_order_and_graph()

        return self.lcm(
            [
                self.get_steps(
                    order, node, graph, lambda node: node[-1] == "Z"
                )
                for node in graph
                if node[-1] == "A"
            ]
        )

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 8")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case _:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()
